hs_composite_resource/models.py

```diff
@@ -80,65 +437,140 @@ def supports_folder_creation(self, folder_full_path):
-        src_file_dir = os.path.dirname(src_full_path)
-
-        def check_directory():
-            path_to_check = ''
-            if istorage.exists(tgt_file_dir):
-                path_to_check = tgt_file_dir
-            else:
-                if tgt_file_dir.startswith(src_file_dir):
-                    path_to_check = src_file_dir
 
-            if path_to_check and not path_to_check.endswith("data/contents"):
-                # it is not the base directory - it must be a directory under base dir
-                res_file_objs = [res_file_obj for res_file_obj in self.files.all() if
-                                 res_file_obj.dir_path == path_to_check]
+        # need to find out which of the following actions the user is trying to do:
+        # renaming a file
+        # renaming a folder
+        # moving a file
+        # moving a folder
+        is_renaming_file = False
+        is_moving_file = False
+        is_moving_folder = False
+
+        tgt_folder, tgt_file_name = os.path.split(tgt_full_path)
+        _, tgt_ext = os.path.splitext(tgt_file_name)
+        if tgt_ext:
+            tgt_file_dir = os.path.dirname(tgt_full_path)
+        else:
+            tgt_file_dir = tgt_full_path
 
-                for res_file_obj in res_file_objs:
-                    if not res_file_obj.logical_file.supports_resource_file_rename or \
-                            not res_file_obj.logical_file.supports_resource_file_move:
-                        return False
+        src_folder, src_file_name = os.path.split(src_full_path)
+        _, src_ext = os.path.splitext(src_file_name)
+        if src_ext:
+            src_file_dir = os.path.dirname(src_full_path)
+        else:
+            src_file_dir = src_full_path
+
+        if src_ext and tgt_ext:
+            is_renaming_file = True
+        elif src_ext:
+            is_moving_file = True
+        elif not istorage.exists(tgt_file_dir):
+            # renaming folder - no restriction
             return True
-
-        res_file_objs = [res_file_obj for res_file_obj in self.files.all() if
-                         res_file_obj.full_path == src_full_path]
-
-        if res_file_objs:
-            res_file_obj = res_file_objs[0]
-            # src_full_path contains file name
-            if not res_file_obj.logical_file.supports_resource_file_rename or \
-                    not res_file_obj.logical_file.supports_resource_file_move:
-                return False
-
-            # check if the target directory allows stuff to be moved there
-            return check_directory()
         else:
-            # src_full_path is a folder path without file name
-            # tgt_full_path also must be a folder path without file name
-            # check that if the target folder contains any files and if any of those files
-            # allow moving stuff there
-            return check_directory()
+            is_moving_folder = True
+
+        if is_renaming_file:
```

Comment:
I'm wondering if we can simplify some of this.  I don't have an idea right now, but making this comment to bring attention to it.  is_renaming_file and is_moving_file have nearly the same logic.

Reply:
Will refactor.